File: MatlabCode/03/IC_singlerun_MC.m
%% File name: IC_singlerun_MC.m

function [x, u, gt] = IC_singlerun_MC(T_max)



%% Parameters

% Maximum capacity
C=10;

% Demand distribution: w = 0,...,W_max
W_max = 4;
pW = [0.2 0.2 0.3 0.2 0.1]; % probabilities

% Stage cost
c1 = 10; % fixed cost for ordering
c2 = 2;  % unitary cost for holding stock

% Reorder if x below x_min
x_min=W_max;

% Markov chain state, possible values 0,1,...,C
XX = [0:C];

% Transition probability matrix
pWf=fliplr(pW); % flip probabilty vector
P1 = [zeros(x_min,C-x_min) repmat(pWf,x_min,1)];
for k=1:C-x_min+1
  P2(k,:) = [zeros(1,k-1) pWf zeros(1,C-x_min-k+1)];
end
P=[P1; P2];     



%% Initialization

%%%%%%%%%%%%%%%%%%%%%%%% Initialization %%%%%%%%%%%%%%%%%%%%

% Inizialize vectors to store results of simulation 
% Not necessary, but improves efficiency

% State 
x=zeros(T_max+1,1);

% Stage cost
gt = zeros(T_max,1);

% Input
u = zeros(T_max,1);



%% Main loop

%%%%%%%%%%%%%%%%%%%%%%%%%%%% Main loop %%%%%%%%%%%%%%%%%%%%%
%
% Initial inventory level (full)
x(1) = C;
for t=1:T_max

  % Save costs and inputs
  %
  % Policy: reorder (up to C) if the level is below x_min
  if x(t) < x_min
    
    % Refill
    u(t) = C - x(t);
    
    % stage cost when ordering
    gt(t) = c1 + c2*x(t);
    
  else
    
    % Do not order
    u(t) = 0;
    
    % stage cost when not ordering
    gt(t) = c2*x(t);
    
  end

  % Select pmf of next state
  %
  % ATTENTION: 
  %                x(t) = 0,1...,C
  %   index of matrix P = 1,2,..,C+1
  %
  %   if x(t)=i -> take row i+1 of matrix P
  %
  r = P(x(t)+1,:);
  
  % Sample next state
  x(t+1) = randsample(XX, 1, true, r);
 
end

File: PythonCode/03/IC_singlerun_MC.py
"""
The function takes the time horizon T_max and the max capacity C_max

returns the simulated states x, orders u, stage costs gt and demands W_real
DATA ARE RETURNED AS COLUMN VECTORS
"""

import numpy as np

def IC_singlerun_MC(T_max:int):

    C_max = 10                                          # max capacity of the storage
    W_max = 4                                           # max demand
    pW = np.array([0.2, 0.2, 0.3, 0.2, 0.1])            # vector of  probabilities
    x_min = 4                                           # minimum number of elements in stock before reordering
    c1 = 10                                             # fixed cost for reordering
    c2 = 2                                              # unitary stage cost

    # transition probability matrix
    pWf = np.flip(pW)                                   # flips the probability vector
    P = np.zeros((C_max + 1, C_max + 1))                # initialize the transition probability matrix
    for i in range(C_max + 1):
        if i < x_min:
            P[i, 6:] = pWf
        else:
            P[i, i-4:i+1] = pWf                         # for readability (i-4 results from i-x_min
                                                        # and i+1 from i-x_min+len(pWf)
    # preallocating vectors
    x = np.zeros(T_max + 1, dtype=int)                  # state (elements in stock)
    gt = np.zeros(T_max, dtype=int)                     # stage cost
    u = np.zeros(T_max, dtype=int)                      # input (elements ordered)

    x[0] = C_max                                        # initial condition

    for t in range(T_max):
        if x[t] < x_min:                                # reorder
            u[t] = C_max - x[t]                         # number of elements to reorder
            gt[t] = c1 + c2*x[t]                        # cost of the order + stocking
        else:                                           # no reorder
            u[t] = 0
            gt[t] = c2*x[t]

        # select the pmf of the next state
        # ATTENTION:    x(t) = 0,1,2,...,C_max
        #               index of matrix P = 0,1,2,...,C_max
        #               if x(t)=i -> take row i of matrix P
        r = P[x[t], :]

        # sample next state
        x[t+1] = np.random.choice(np.arange(C_max + 1, dtype=int), 1, replace=True, p=r)

    # returned as row vectors
    return x.reshape((-1, 1)), u.reshape((-1, 1)), gt.reshape((-1, 1))
